nameko_sqlalchemy/database_session.py

```diff
@@ -22,7 +30,14 @@ def setup(self):
             'service_name': service_name,
             'declarative_base_name': decl_base_name,
         })
-        self.engine = create_engine(self.db_uri)
+
+        self.session_options.update(
+            self.container.config.get(DB_SESSION_OPTIONS_KEY, {}))
+        self.engine_options.update(
+            self.container.config.get(DB_ENGINE_OPTIONS_KEY, {}))
```

Comment:
I think it would be a nicer API if the instance-specified option overrode any config-specified option. That is what we do in Nameko with the messaging options that can also be specified in config:

https://github.com/nameko/nameko/blob/f496349d4caf368d69a6d433205bd28e3058e250/nameko/messaging.py#L148-L157

https://github.com/nameko/nameko/blob/f496349d4caf368d69a6d433205bd28e3058e250/nameko/messaging.py#L167

Reply:
Not sure why those permalinks haven't auto-expanded out on the preview. Click through to see what I mean

Reply:
Not sure if the codebase setup should override setup from config. Normally I would expect that config entries have the last say. But I understand that there is a use case of having multiple instances of the same dependency, each with different setup.

I'm thinking of removing loading the engine and session options from config in the DP in favor of setting them up when instantiating extensions on the service level using a [global config](https://github.com/nameko/nameko/pull/520).

Passing to constructor options loaded from config (by users) also eliminates discussions (extension authors) about [how the config entries should be structured](https://github.com/onefinestay/nameko-sqlalchemy/pull/14#issuecomment-338707726).

Reply:
Examples of using the [proposed nameko config](https://github.com/nameko/nameko/pull/520):
```
from nameko import config
from nameko_sqlalchemy import Database

class Service:
    db = Database(engine_options=config.DB_ENGINE_OPTIONS)
```
User governs the config structure:
``` 
from nameko import config
from nameko_sqlalchemy import Database

class Service:
    db = Database(engine_options={'pool_size': config.DB_POOL_SIZE})
```